src/Gaufrette/Adapter/Ftp.php

```diff
@@ -366,12 +378,36 @@ private function isDir($directory)
             return true;
         }
 
-        if (!@ftp_chdir($this->getConnection(), $directory)) {
-            return false;
+        $chDirResult = false;
+        try {
+            $chDirResult = ftp_chdir($this->getConnection(), $this->directory);
+
+            // change directory again to return in the base directory
+            ftp_chdir($this->getConnection(), $this->directory);
+        } catch(\Exception $e) {
+
+            // Cache the actual value of the passive mode
+            $actualPassiveMode  = $this->passive;
```

Comment:
Actually, I don't think you have to change this flag. It should be RO, and it is only used to specify whether the backend supports passive mode. See its usage [1] for more details. Changing the value of the flag do not brings any change in the behavior of the PHP's FTP functions. The flag is only used as an indicator in this adapter.

So in the end, the call to `is_dir` should be made only when the passive mode is enabled. When it is not enabled, you can throw the `$e` exception as we can't use `is_dir` :

> If the server does not support passive mode ftp, the connection will fail. 

*[2]*

Also, it could be good to add a comment with the link to [2] above the usage of `is_dir` and the check of the passive mode to ease understanding.

[1] https://github.com/KnpLabs/Gaufrette/blob/350f0f68e72c9f39845413c9b775ac49ffa18e8b/src/Gaufrette/Adapter/Ftp.php#L548
[2] https://php.net/manual/en/wrappers.ftp.php